generic3g/specs/VariableSpec_private.F90

```diff
@@ -0,0 +1,216 @@
+
+   logical function is_in_realR4(bounds, t) result(lval)
+      real(kind=ESMF_KIND_R4), intent(in) :: bounds(:)
+      real(kind=ESMF_KIND_R4), intent(in) :: t
+      integer :: i
+
+      lval = .FALSE.
+      if(size(bounds) < 2) return
+
+      do i = 2, mod(size(bounds), 2), 2
+         lval = t >= bounds(i-1) .and. t <= bounds(i)
+         if(lval) exit
+      end do
+
+   end function is_in_realR4
+
+   logical function is_not_empty(string)
+      character(len=*), intent(in) :: string
+
+      is_not_empty = len_trim(string) > 0
+
+   end function is_not_empty
+
+   logical function no_test(v)
+      class(*), intent(in) :: v
+
+      no_test = .TRUE.
+
+   end function no_test
+
+   logical function string_in_vector(string, vector) result(in_vector)
+      character(len=*), intent(in) :: string
+      class(StringVector), intent(in) :: vector
+      type(StringVectorIterator) :: e, iter
+
+      in_vector = .TRUE.
+      e = vector%end()
+      iter = vector%begin()
+      do while(iter /= e)
+         if(string == iter%of()) return
+         call iter%next()
+      end do
+      in_vector = .FALSE.
+
+   end function string_in_vector
+
+   logical function is_vector_in_string_vector(V0, V) result(lval)
+      class(StringVector), intent(in) :: V0
+      class(StringVector), intent(in) :: V
+      type(StringVectorIterator) :: iter, e
+
+      lval = .FALSE.
+      iter = V%begin()
+      e = V%end()
+      do while(iter /= e)
+         if(.not. string_in_vector(iter%of(), V0)) return
+         call iter%next()
+      end do
+      lval = .TRUE.
+
+   end function is_vector_in_string_vector
+
+   subroutine validate_short_name(v, rc)
+      character(len=*), intent(in) :: v
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      _ASSERT(is_valid_identifier(v), 'Invalid value')
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_short_name
+
+   subroutine validate_regrid(param, method, rc)
+      type(EsmfRegridderParam), optional, intent(in) :: param
+      type(ESMF_RegridMethod_Flag), optional, intent(in) :: method
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      if(present(param)) then
+         _ASSERT(.not. present(method), 'regrid_param and regrid_method are mutually exclusive.')
+      end if
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_regrid
+
+   subroutine validate_state_intent(v, rc)
+      type(ESMF_StateIntent_Flag), intent(in) :: v
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      _ASSERT(valid_state_intent(v), 'Invalid value')
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_state_intent
+
+   subroutine validate_state_item(v, rc)
+      type(ESMF_StateItem_Flag), intent(in) :: v
+      integer, optional, intent(out) :: rc
+      integer :: status
+
+      _ASSERT(valid_state_item(v), 'Invalid value')
+      _RETURN(_SUCCESS)
+
+   end subroutine validate_state_item
+
+   logical function valid_state_intent(val)
```

Comment:
A thought:   Maybe we want to overload a generic name `is_valid()` with a largish set of overloads:
```
is_valid_state_intent()
is_valid_state_item()  ! hate the esmf name here
...
```

Reply:
Because some of the members have the same type (short_name and expression), I decided against a generic name.